### ISSUE TITLE
cmd/gc: incorrect nosplit stack calculation

### ISSUE THREAD

**Issue by dvyukov** [golang/go#10079]
https://go-review.googlesource.com/#/c/3742/ failed noopt builder:

```
runtime.atomicstorep: nosplit stack overflow
	504	assumed on entry to runtime.atomicstorep
	472	after runtime.atomicstorep uses 32
	464	on entry to runtime.writebarrierptr_noshadow
	424	after runtime.writebarrierptr_noshadow uses 40
	416	on entry to runtime.shadowptr
	384	after runtime.shadowptr uses 32
	376	on entry to runtime.inheap
	512	after runtime.inheap uses -136
	504	on entry to runtime.panicindex
	512	after runtime.panicindex uses -8
	504	on entry to runtime.gopanic
	512	after runtime.gopanic uses -8
	504	on entry to runtime.printany
	512	after runtime.printany uses -8
	504	on entry to runtime.assertE2I2
	512	after runtime.assertE2I2 uses -8
	504	on entry to runtime.getitab
	512	after runtime.getitab uses -8
	504	on entry to runtime.newobject
	512	after runtime.newobject uses -8
	504	on entry to runtime.mallocgc
	512	after runtime.mallocgc uses -8
	504	on entry to runtime.startGC
	512	after runtime.startGC uses -8
	504	on entry to runtime.gc
	512	after runtime.gc uses -8
	504	on entry to runtime.semacquire
	512	after runtime.semacquire uses -8
	504	on entry to runtime.acquireSudog
	-48	after runtime.acquireSudog uses 552
runtime.chansend1: nosplit stack overflow
	504	assumed on entry to runtime.chansend1
	448	after runtime.chansend1 uses 56
	440	on entry to runtime.chansend
	512	after runtime.chansend uses -72
	504	on entry to runtime.releaseSudog
	64	after runtime.releaseSudog uses 440
	56	on entry to runtime.throw
	24	after runtime.throw uses 32
	16	on entry to runtime.startpanic
	8	after runtime.startpanic uses 8
	0	on entry to runtime.systemstack
	-8	on entry to function pointer
	-16	on entry to runtime.morestack
```

http://build.golang.org/log/549c56185ba7b762a995a6aa71f4f865c14286e1

But that calculation looks incorrect for several reasons:
1. Functions can not consume negative amount of stack
2. runtime.mallocgc is not a nosplit function, so it should break the sequence
3. It must break all tests, because all tests send to channel, but it broke only runtime test

@rsc

**Comment by bradfitz** [comment]
Aside: I don't have the noopt builder as part of the trybot set. We could, though, and probably should because this has come up now a number of times. /cc @adg

**Comment by eliasnaur** [comment]
For fun and profit I played with this a bit. I found that
 
 - The builder runs with -N (disable optimizations) and -l (disable inlining). Only -N is necessary to reproduce the error.
 - The negative amounts of stack corresponds to splitting functions resetting the available stack space. I verified that every negative amount corresponds to a splitting function. As corroborating evidence, every negative amount also corresponds to a reset to 512, presumably the stack limit.

I don't know why the error only appears in the runtime test, but apart from that the error seems benign in the sense that -N disables regopt which in turn increases stack pressure, triggering a stack overflow. An obvious fix is to raise the stack limit for -N builds, but I'm not sure that is easily done across the compilation/linking split.
